internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,278 @@
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+
+	// Deletion is equal to 0 desired replicas.
+	if !ms.DeletionTimestamp.IsZero() {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionTrue,
+			Reason:  clusterv1.MachineSetScalingDownV1Beta2Reason,
+			Message: "MachineSet is getting deleted",
+		})
+		return
+	}
+
+	// Scaling down.
+	if int32(len(machines)) > (desiredReplicas) {
+		messages := []string{fmt.Sprintf("ScalingDown from %d to %d replicas", len(machines), desiredReplicas)}
+
+		if len(machinesStaleDeleting) > 0 {
+			messages = append(messages, fmt.Sprintf("Machines stuck in deletion for more than 30 minutes: %s", clog.StringListToString(machinesStaleDeleting)))
+		}
+
+		// Scaling is blocked when the bootstrap or infrastructure objects do not exist.
+		if bootstrapObjectNotFound || infrastructureObjectNotFound {
+			messages[0] = fmt.Sprintf("%s is blocked because %s", messages[0], calculateMissingReferencesMessage(bootstrapObjectNotFound, infrastructureObjectNotFound))
+		}
+
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionTrue,
+			Reason:  clusterv1.MachineSetScalingDownV1Beta2Reason,
+			Message: strings.Join(messages, ", "),
+		})
+		return
+	}
+
+	// Not scaling down.
+	var message string
+	// The referenced objects are required to exist for the current MachineSet.
+	// Add a message if that's not the case.
+	if isCurrentMachineSet(ms, owner) && (bootstrapObjectNotFound || infrastructureObjectNotFound) {
+		message = fmt.Sprintf("ScalingDown would fail because %s", calculateMissingReferencesMessage(bootstrapObjectNotFound, infrastructureObjectNotFound))
+	}
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+		Status:  metav1.ConditionFalse,
+		Reason:  clusterv1.MachineSetNotScalingDownV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func setMachinesReadyCondition(machineSet *clusterv1.MachineSet, machines []*clusterv1.Machine) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(machineSet, metav1.Condition{
+			Type:    clusterv1.MachineSetMachinesReadyV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetMachinesReadyV1Beta2Condition,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	if len(machines) == 0 {
+		v1beta2conditions.Set(machineSet, metav1.Condition{
+			Type:   clusterv1.MachineSetMachinesReadyV1Beta2Condition,
+			Status: metav1.ConditionTrue,
```

Comment:
Should this be unknown?

Reply:
Good question.

I'd say no because 0 replicas is something a MachineSet may be totally be valid to be in. If one aggregates over all machinesets to see if one of them / its machines is in a bad state, he just needs to watch out if any of them is not true (unknown or false)

Reply:
yep, 0 machine is ok.
but MachinesReady = true when there are no machines seems border line.

Reply:
I think it should be true. There is nothing unknown about this situation